python/ql/lib/semmle/python/frameworks/Bottle.qll

```diff
@@ -0,0 +1,179 @@
+/**
+ * Provides classes modeling security-relevant aspects of the `bottle` PyPI package.
+ * See https://bottlepy.org/docs/dev/.
+ */
+
+private import python
+private import semmle.python.Concepts
+private import semmle.python.ApiGraphs
+private import semmle.python.dataflow.new.RemoteFlowSources
+private import semmle.python.frameworks.internal.InstanceTaintStepsHelper
+private import semmle.python.frameworks.Stdlib
+
+/**
+ * INTERNAL: Do not use.
+ *
+ * Provides models for the `bottle` PyPI package.
+ * See https://bottlepy.org/docs/dev/.
+ */
+module Bottle {
+  /** Gets a reference to the `bottle` module. */
+  API::Node bottle() { result = API::moduleImport("bottle") }
+
+  /** Provides models for the `bottle` module. */
+  module BottleModule {
+    /**
+     * Provides models for Bottle applications.
+     */
+    module App {
+      /** Gets class `bottle.Bottle`) */
+      API::Node cls() { result = API::moduleImport("bottle").getMember("Bottle") }
+
+      /** Gets a reference to a Bottle application (an instance of `bottle.Bottle`) */
+      API::Node instance() { result = cls().getReturn() }
+
+      /** Gets a reference to a Bottle application (an instance of `bottle.app`) */
+      API::Node app() { result = bottle().getMember("app").getReturn() }
+    }
+
+    /** Provides models for functions that are possible "views" */
+    module View {
+      /**
+       * A Bottle view callable, that handles incoming requests.
+       */
+      class ViewCallable extends Function {
+        ViewCallable() { this = any(BottleRouteSetup rs).getARequestHandler() }
+      }
+
+      private class BottleRouteSetup extends Http::Server::RouteSetup::Range, DataFlow::CallCfgNode {
+        BottleRouteSetup() {
+          this =
+            [
+              App::instance()
+                  .getMember(["route", "get", "post", "put", "delete", "patch"])
+                  .getACall(),
+              App::app().getMember(["route", "get", "post", "put", "delete", "patch"]).getACall(),
+              bottle().getMember(["route", "get", "post", "put", "delete", "patch"]).getACall()
+            ]
+        }
+
+        override DataFlow::Node getUrlPatternArg() {
+          result in [this.getArg(0), this.getArgByName("route")]
+        }
+
+        override string getFramework() { result = "Bottle" }
+
+        override Parameter getARoutedParameter() { none() }
+
+        override Function getARequestHandler() { result.getADecorator().getAFlowNode() = node }
+      }
+    }
+
+    /** Provides models for the `bottle.response` module */
+    module Response {
+      /** Gets a reference to the `bottle.response` module. */
+      API::Node response() { result = bottle().getMember("response") }
```

Comment:
Shoudl we include `bottle.Response.getReturn()` (with capital R) here? See https://bottlepy.org/docs/dev/api.html#the-response-object.

Reply:
I'm modeling https://bottlepy.org/docs/dev/api.html#bottle.response which is global and not a function. Don't think its necessary.

Reply:
I think it is an additional may to obtain/modify responses. It may only see modest use ([partial ci search](https://github.com/search?type=code&q=%2F%28%3F-i%29bottle%5C.Response%2F+language%3APython)), but it should be easy enough for us to recognise it?